torch/utils/data/standard_pipes.ipynb

```diff
@@ -0,0 +1,1006 @@
+      "\u001b[0;31mException\u001b[0m: ('Failed to group items', '[(0, 0)]')"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(15).map(lambda x: (x % 3, x)).shuffle().groupby(lambda x: x[0], guaranteed_group_size = 2, buffer_size = 6)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "To avoid this error and drop incomplete groups, use `drop_remaining` argument"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 47,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[(2, 5), (2, 11), (2, 14), (2, 2)]\n",
+      "[(1, 7), (1, 10), (1, 13), (1, 4), (1, 1)]\n",
+      "[(0, 9), (0, 0), (0, 6), (0, 3), (0, 12)]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(15).map(lambda x: (x % 3, x)).shuffle().groupby(lambda x: x[0], guaranteed_group_size = 2, buffer_size = 6, drop_remaining = True)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "## Zip\n",
+    "\n",
+    "Function: `zip`\n",
+    "\n",
+    "Description: \n",
+    "\n",
+    "Alternatives:\n",
+    "\n",
+    "Arguments:\n",
+    " \n",
+    "Example:"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 48,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "(0, 0)\n(1, 3)\n(2, 1)\n(3, 2)\n(4, 4)\n"
+     ]
+    }
+   ],
+   "source": [
+    "_dp = ExampleIterPipe(5).shuffle()\n",
+    "dp = ExampleIterPipe(5).zip(_dp)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "## Fork\n",
```

Comment:
What happens to Fork if only one of the iterators is being consumed? It has to buffer data somewhere, is there an ability to control the size of the buffer somehow? Or it'd just OOM?

and what's the typical use case for? just curious whether we need it at all

Reply:
Current hack contains buffer, real implementation will be buffer based + wait consumers to process items before reading more from source. 

Classic use-case is (pseudocode):
```
files_dp = ReadFilesFromTar()
images_dp, labels_dp = files_dp.fork(2)
images_dp = images_dp.filter(lambda x: x[0].endswith('.jpg')).map(images_decoder())
labels_dp = labels_dp.filter(lambda x: x[0].endswith('.json')).map(json_decoder())
combined_dp = datapipes.iter.concat_greedy(images_dp, labels_dp).groupby(...)
```